src/ucs/memory/memtype_cache.c

```diff
@@ -300,6 +315,30 @@ UCS_PROFILE_FUNC(ucs_status_t, ucs_memtype_cache_lookup,
     return status;
 }
 
+int ucs_memtype_cache_is_empty()
+{
+    return (ucs_memtype_cache_global_instance &&
+            !ucs_memtype_cache_global_instance->pgtable.num_regions);
+}
+
+static ucs_status_t ucs_memtype_cache_init_handler()
+{
+    ucs_status_t status;
+
+    status = ucm_set_event_handler(UCM_EVENT_MEM_TYPE_ALLOC |
+                                   UCM_EVENT_MEM_TYPE_FREE |
+                                   UCM_EVENT_FLAG_EXISTING_ALLOC,
+                                   1000, ucs_memtype_cache_event_callback,
+                                   ucs_memtype_cache_global_instance);
```

Comment:
what if we do not set ucs_memtype_cache_global_instance variable at this point yet---
instead, we create local memtype_cache in ucs_memtype_cache_get_global. then, pass it to ucs_memtype_cache_init_handler(). ucs_memtype_cache_event_callback function will **have** to use "arg" instead to the global instance and also pass it to update_internal.
then, if all is good, ucs_memtype_cache_get_global will set the global instance to this local memtype_cache var.

this way - by using "arg" parameter of memtype event - can move this memtype set handler back to memtype_cache constructor.